![tracker](https://us-central1-vertex-ai-mlops-369716.cloudfunctions.net/pixel-tracking?path=statmike%2Fvertex-ai-mlops%2Fdata%2Bai%2Fbq-ml%2Ffunctions%2Fimage&file=image.ipynb)
<!--- header table --->
<table align="left">
<tr>
  <td style="text-align: center">
    <a href="https://github.com/statmike/vertex-ai-mlops/blob/main/data%2Bai/bq-ml/functions/image/image.ipynb">
      <img width="32px" src="https://www.svgrepo.com/download/217753/github.svg" alt="GitHub logo">
      <br>View on<br>GitHub
    </a>
  </td>
  <td style="text-align: center">
    <a href="https://colab.research.google.com/github/statmike/vertex-ai-mlops/blob/main/data%2Bai/bq-ml/functions/image/image.ipynb">
      <img width="32px" src="https://www.gstatic.com/pantheon/images/bigquery/welcome_page/colab-logo.svg" alt="Google Colaboratory logo">
      <br>Run in<br>Colab
    </a>
  </td>
  <td style="text-align: center">
    <a href="https://console.cloud.google.com/vertex-ai/colab/import/https%3A%2F%2Fraw.githubusercontent.com%2Fstatmike%2Fvertex-ai-mlops%2Fmain%2Fdata%252Bai%2Fbq-ml%2Ffunctions%2Fimage%2Fimage.ipynb">
      <img width="32px" src="https://lh3.googleusercontent.com/JmcxdQi-qOpctIvWKgPtrzZdJJK-J3sWE1RsfjZNwshCFgE_9fULcNpuXYTilIR2hjwN" alt="Google Cloud Colab Enterprise logo">
      <br>Run in<br>Colab Enterprise
    </a>
  </td>      
  <td style="text-align: center">
    <a href="https://console.cloud.google.com/bigquery/import?url=https://github.com/statmike/vertex-ai-mlops/blob/main/data%2Bai/bq-ml/functions/image/image.ipynb">
      <img width="32px" src="https://www.gstatic.com/images/branding/gcpiconscolors/bigquery/v1/32px.svg" alt="BigQuery logo">
      <br>Open in<br>BigQuery Studio
    </a>
  </td>
  <td style="text-align: center">
    <a href="https://console.cloud.google.com/vertex-ai/workbench/deploy-notebook?download_url=https://raw.githubusercontent.com/statmike/vertex-ai-mlops/main/data%2Bai/bq-ml/functions/image/image.ipynb">
      <img width="32px" src="https://www.gstatic.com/images/branding/gcpiconscolors/vertexai/v1/32px.svg" alt="Vertex AI logo">
      <br>Open in<br>Vertex AI Workbench
    </a>
  </td>
</tr>
<tr>
  <td colspan="5" style="text-align: right">
    <b>Share This On: </b> 
    <a href="https://www.linkedin.com/sharing/share-offsite/?url=https://github.com/statmike/vertex-ai-mlops/blob/main/data%252Bai/bq-ml/functions/image/image.ipynb"><img src="https://upload.wikimedia.org/wikipedia/commons/8/81/LinkedIn_icon.svg" alt="Linkedin Logo" width="20px"></a> 
    <a href="https://reddit.com/submit?url=https://github.com/statmike/vertex-ai-mlops/blob/main/data%252Bai/bq-ml/functions/image/image.ipynb"><img src="https://redditinc.com/hubfs/Reddit%20Inc/Brand/Reddit_Logo.png" alt="Reddit Logo" width="20px"></a> 
    <a href="https://bsky.app/intent/compose?text=https://github.com/statmike/vertex-ai-mlops/blob/main/data%252Bai/bq-ml/functions/image/image.ipynb"><img src="https://upload.wikimedia.org/wikipedia/commons/7/7a/Bluesky_Logo.svg" alt="BlueSky Logo" width="20px"></a> 
    <a href="https://twitter.com/intent/tweet?url=https://github.com/statmike/vertex-ai-mlops/blob/main/data%252Bai/bq-ml/functions/image/image.ipynb"><img src="https://upload.wikimedia.org/wikipedia/commons/5/5a/X_icon_2.svg" alt="X (Twitter) Logo" width="20px"></a> 
  </td>
</tr>
<tr>
  <td colspan="5" style="text-align: right">
    <b>Connect With Author On: </b> 
    <a href="https://www.linkedin.com/in/statmike"><img src="https://upload.wikimedia.org/wikipedia/commons/8/81/LinkedIn_icon.svg" alt="Linkedin Logo" width="20px"></a>
    <a href="https://www.github.com/statmike"><img src="https://www.svgrepo.com/download/217753/github.svg" alt="GitHub Logo" width="20px"></a> 
    <a href="https://www.youtube.com/@statmike-channel"><img src="https://upload.wikimedia.org/wikipedia/commons/f/fd/YouTube_full-color_icon_%282024%29.svg" alt="YouTube Logo" width="20px"></a>
    <a href="https://bsky.app/profile/statmike.bsky.social"><img src="https://upload.wikimedia.org/wikipedia/commons/7/7a/Bluesky_Logo.svg" alt="BlueSky Logo" width="20px"></a> 
    <a href="https://x.com/statmike"><img src="https://upload.wikimedia.org/wikipedia/commons/5/5a/X_icon_2.svg" alt="X (Twitter) Logo" width="20px"></a>
  </td>
</tr>
<tr>
  <td colspan="5" style="text-align: right">
    <a href="https://raw.githubusercontent.com/statmike/vertex-ai-mlops/main/data%2Bai/bq-ml/functions/image/image.ipynb"><img src="https://www.svgrepo.com/download/5445/download-button.svg" alt="Download icon" width="20px"></a> <a href="https://raw.githubusercontent.com/statmike/vertex-ai-mlops/main/data%2Bai/bq-ml/functions/image/image.ipynb">Download Notebook</a> <i>(right-click and "Save As")</i>
  </td>
</tr>
</table>


# Image Preprocessing — `ML.DECODE_IMAGE`, `ML.RESIZE_IMAGE`, `ML.CONVERT_COLOR_SPACE`, `ML.CONVERT_IMAGE_TYPE`

Four model-free scalar functions that turn image **bytes** into the numeric `STRUCT` a vision model consumes: decode, resize, change color space, quantize to integers. They train nothing, create no model object, and nest freely inside one another and inside a `TRANSFORM` clause.

They are also the least accurately documented family in this project, which is most of what this notebook is for. Everything below is measured here, against images this notebook builds pixel by pixel, so every claim has a cell you can read.

> **GOTCHA — these four functions require a BigQuery Editions reservation.** On-demand (per-byte) pricing fails outright. Setup demonstrates the exact error, then creates a small autoscaling `ENTERPRISE` reservation and deletes it in Cleanup. **None of the four official reference pages mentions this.** `MATRIX_FACTORIZATION` ([`models/matrix_factorization/`](../../models/matrix_factorization/)) is the only other thing in this project with the same requirement.

**What this notebook establishes, all measured here rather than quoted:**

1. The output `STRUCT`'s fields are named **`dimensions`** and **`values`** — not `shape`.
2. `ML.DECODE_IMAGE` takes **any** `BYTES` value, not only an object table's `data` pseudocolumn. That is what makes the rest of this notebook reproducible without a bucket.
3. **Decoding always returns 3 channels**, whatever the source had — an RGBA image's alpha is silently dropped (not composited), and a single-channel grayscale image is replicated into three identical channels. `ML.CONVERT_COLOR_SPACE(..., 'GRAYSCALE')` then takes it back down to **1**, so the channel count is set by the last function in the chain, not by the image.
4. The `[0, 1)` and `[0, 255)` ranges printed on the reference pages are wrong at the top end — a pure-white pixel is exactly `1.0`, and `ML.CONVERT_IMAGE_TYPE` returns exactly `255`.
5. `ML.CONVERT_IMAGE_TYPE` is **one-way**: there is no integer→float signature.
6. `GRAYSCALE` and `YIQ`'s Y channel are **two different lumas**, differing in the fourth decimal place.
7. `ML.RESIZE_IMAGE` is **type-preserving and float32**, while decode and color-space conversion are float64-exact.
8. A `preserve_aspect_ratio = TRUE` fit that rounds a dimension to zero raises an **internal error the client library retries for minutes** unless you turn job retry off.
9. All four are **TensorFlow image ops**. Diffed against TensorFlow directly: `HSV`, bilinear resize and the integer conversion are **bit-exact**, `YIQ`/`YUV` match to a last-place unit of a double, and the one row that misses at single precision misses because **`tf.image` computes it in float32 while BigQuery uses double**.

**When to use these:** you have image bytes in Cloud Storage and a vision model — [imported](../../models/imported/) or [remote](../../models/remote/) — that expects a fixed-size numeric tensor. These four get you from one to the other in SQL, with no data leaving BigQuery. For *generative* work on images (captioning, classification by prompt, embeddings) the path is entirely different — see **Related content** below.

**Data:** six images this notebook constructs from NumPy arrays — small enough that every pixel value is printable and checkable by hand, which is the only way the precision findings above are provable. They are uploaded to Cloud Storage so Step 8 can exercise the real object-table path.

**Related content:** [`models/imported/`](../../models/imported/) — how the model these features feed gets into BigQuery. [`models/matrix_factorization/`](../../models/matrix_factorization/) — the same reservation pattern, for the same reason. For multimodal generative AI over images (`ML.GENERATE_TEXT`, `AI.GENERATE_TABLE`, `ML.GENERATE_EMBEDDING` over an ObjectRef), see the sibling project [`bq-ai-functions`](../../../bq-ai-functions/) — that is a foundation-model path, not this preprocessing one.

**References:** [Full reference](../../reference/model-free-functions.md) | [ML.DECODE_IMAGE](https://cloud.google.com/bigquery/docs/reference/standard-sql/bigqueryml-syntax-decode-image) | [ML.RESIZE_IMAGE](https://cloud.google.com/bigquery/docs/reference/standard-sql/bigqueryml-syntax-resize-image) | [ML.CONVERT_COLOR_SPACE](https://cloud.google.com/bigquery/docs/reference/standard-sql/bigqueryml-syntax-convert-color-space) | [ML.CONVERT_IMAGE_TYPE](https://cloud.google.com/bigquery/docs/reference/standard-sql/bigqueryml-syntax-convert-image-type) | [Object tables](https://cloud.google.com/bigquery/docs/object-tables) | [Setup guide](../../setup/)


---
## Setup

Set your project, location, and bucket, authenticate, create the shared dataset — then deal with the reservation requirement, which is the part no other `functions/` notebook has.

In [1]:
PROJECT_ID = 'statmike-mlops-349915'  # <-- Replace with your project ID
LOCATION = 'US'  # BigQuery dataset location
DATASET_ID = 'bq_ml'  # Shared dataset across all bq-ml notebooks
BUCKET = 'statmike-mlops-349915'  # <-- Replace with your GCS bucket (same location as DATASET_ID)
CONNECTION_ID = 'bq_ml_image_demo'  # Cloud Resource Connection for the object table in Step 8
RESERVATION_ID = 'bqml-image-temp'  # Temporary reservation name (Setup creates it, Cleanup removes it)


### Environment

> **Already set up the project environment?** The cell below is a no-op — packages are already in your kernel. See the [Setup Reference](../../setup/) for details.
>
> **Running standalone** (Colab, Colab Enterprise, Vertex AI Workbench)? The cell below installs required packages into your current kernel.

In [2]:
import subprocess, sys, shutil

def install(*packages):
    """Install packages using uv (fast) with pip fallback."""
    uv = shutil.which('uv')
    if uv:
        subprocess.check_call([uv, 'pip', 'install', '-q', '--python', sys.executable, *packages])
    else:
        subprocess.check_call([sys.executable, '-m', 'pip', 'install', '-q', '--upgrade', *packages])

install('google-cloud-bigquery', 'google-cloud-bigquery-reservation', 'google-cloud-storage',
        'db-dtypes', 'bigquery-magics', 'bigframes', 'pillow', 'numpy', 'tensorflow')

In [3]:
# Authenticate (Colab only — other environments use Application Default Credentials)
try:
    from google.colab import auth
    auth.authenticate_user()
except ImportError:
    pass  # Not in Colab — ADC is used automatically

In [4]:
from google.cloud import bigquery, storage
import pandas as pd

client = bigquery.Client(project=PROJECT_ID)
gcs_client = storage.Client(project=PROJECT_ID)
gcs_bucket = gcs_client.bucket(BUCKET)
pd.set_option('display.max_colwidth', None)

# Every image query in this notebook runs cache-off. The reservation demo two cells
# below depends on it: a cached result is served without slots and would hide the
# very failure the cell exists to show.
no_cache = bigquery.QueryJobConfig(use_query_cache=False)

# Create the shared dataset (idempotent)
dataset_ref = bigquery.DatasetReference(PROJECT_ID, DATASET_ID)
dataset = bigquery.Dataset(dataset_ref)
dataset.location = LOCATION
client.create_dataset(dataset, exists_ok=True)
print(f'Dataset {PROJECT_ID}.{DATASET_ID} ready')

# Register %%bigquery cell magic (auto-loaded in Colab, needed elsewhere)
%load_ext bigquery_magics

Dataset statmike-mlops-349915.bq_ml ready



### Build the test images

Six images, defined as NumPy arrays so the ground truth is in the notebook rather than in a file. Two pixels carry most of the weight: **(200, 100, 50)**, a warm color with all three channels distinct, and **(0, 128, 255)**, which pins the low end, the middle, and the high end at once.

Each is encoded to PNG (and one to JPEG) in memory. The bytes are kept two ways: base64, so they can be inlined into SQL as a `BYTES` literal, and raw, so Step 8 can upload them to Cloud Storage.

In [5]:
import io, base64
import numpy as np
from PIL import Image

# Ground truth, as arrays. Shape is (height, width, channels).
swatch_a   = np.array([[[200, 100, 50], [0, 128, 255]]], dtype=np.uint8)                        # 1x2 RGB
gradient_a = np.array([[[x * 32, y * 64, 128] for x in range(8)] for y in range(4)], np.uint8)  # 4x8 RGB
alpha_a    = np.array([[[200, 100, 50, 255], [0, 128, 255, 128]]], dtype=np.uint8)              # 1x2 RGBA
white_a    = np.array([[[255, 255, 255]]], dtype=np.uint8)                                      # 1x1 RGB

def encode(array, fmt='PNG', mode=None):
    image = Image.fromarray(array)
    if mode:
        image = image.convert(mode)
    buffer = io.BytesIO()
    image.save(buffer, format=fmt)
    return buffer.getvalue()

raw = {
    'swatch.png':   encode(swatch_a),
    'gradient.png': encode(gradient_a),
    'alpha.png':    encode(alpha_a),
    'gray.png':     encode(swatch_a, mode='L'),   # single-channel PNG
    'white.png':    encode(white_a),
    'swatch.jpg':   encode(swatch_a, fmt='JPEG'),
}
b64 = {name: base64.b64encode(data).decode() for name, data in raw.items()}

def img(name):
    """SQL fragment: this image's bytes as a literal, decoded."""
    return f"ML.DECODE_IMAGE(FROM_BASE64('{b64[name]}'))"

for name, data in raw.items():
    print(f'{name:14s} {len(data):5d} bytes')

swatch.png        72 bytes
gradient.png      81 bytes
alpha.png         74 bytes
gray.png          68 bytes
white.png         69 bytes
swatch.jpg       652 bytes



### The undocumented prerequisite: a slot reservation

The cell below runs the simplest possible `ML.DECODE_IMAGE` call under whatever pricing model this project currently has. On a project with no reservation, it fails.

In [6]:
from google.api_core.exceptions import BadRequest

probe = f'SELECT {img("swatch.png")}.dimensions AS dims'
try:
    print('On-demand result:', list(client.query(probe, job_config=no_cache).result())[0]['dims'])
    print('\nThis project already has a reservation assigned — the next cell will reuse or create one anyway.')
except BadRequest as e:
    print('On-demand FAILED, as expected:\n')
    print(str(e).split(';')[0])

On-demand FAILED, as expected:

400 BigQuery's ML image-processing functions requires reservation, but no reservation was assigned for job type `QUERY`, to project `statmike-mlops-349915` or its parent, in location `US`.


That message names the requirement precisely — `job type QUERY`, this project or its parent, this location — and it appears on none of the four function reference pages, nor on the [manual preprocessing](https://cloud.google.com/bigquery/docs/manual-preprocessing) or [object table inference](https://cloud.google.com/bigquery/docs/object-table-inference) pages that tell you to use these functions.

The fix is a BigQuery Editions reservation. This notebook creates a small **autoscaling `ENTERPRISE`** one — 0 baseline slots, scaling to 100 only while a query actually runs — and deletes it in Cleanup.

> **This is not a capacity commitment.** No upfront purchase, no minimum term. It bills per-second for slot-time actually consumed and idles at zero cost. Do not substitute a commitment for it.

**Idempotent:** re-running detects and reuses an existing reservation/assignment instead of erroring or duplicating.

> **Assignment propagation is not atomic, so this cell waits on evidence rather than on a clock.** A probe query can succeed and the very next query — seconds later, same session, same project — fail again with the identical "no reservation was assigned" error. That rules out both a fixed sleep and a single successful probe as evidence, so the loop below requires several consecutive successes before continuing. How long it takes is not fixed either: watch the elapsed times it prints.

In [7]:
import time
from google.cloud import bigquery_reservation_v1
from google.api_core.exceptions import AlreadyExists

reservation_client = bigquery_reservation_v1.ReservationServiceClient()
reservation_parent = f'projects/{PROJECT_ID}/locations/{LOCATION}'
reservation_name = f'{reservation_parent}/reservations/{RESERVATION_ID}'

try:
    reservation = reservation_client.create_reservation(
        parent=reservation_parent,
        reservation_id=RESERVATION_ID,
        reservation=bigquery_reservation_v1.Reservation(
            slot_capacity=0,
            edition=bigquery_reservation_v1.Edition.ENTERPRISE,
            autoscale=bigquery_reservation_v1.Reservation.Autoscale(max_slots=100),
            ignore_idle_slots=True,
        ),
    )
    print(f'Created {reservation.name}')
except AlreadyExists:
    reservation = reservation_client.get_reservation(name=reservation_name)
    print(f'Reservation already exists — reusing {reservation.name}')

existing_assignments = list(reservation_client.list_assignments(parent=reservation.name))
if existing_assignments:
    assignment = existing_assignments[0]
    print(f'Assignment already exists — reusing {assignment.name}')
else:
    assignment = reservation_client.create_assignment(
        parent=reservation.name,
        assignment=bigquery_reservation_v1.Assignment(
            job_type=bigquery_reservation_v1.Assignment.JobType.QUERY,
            assignee=f'projects/{PROJECT_ID}',
        ),
    )
    print(f'Created {assignment.name}')

# Wait for the reservation to answer CONSISTENTLY, not just once. A fixed sleep
# is not enough: one probe succeeded here and the next one failed.
required_streak, streak, waited = 3, 0, 0
while streak < required_streak:
    try:
        dims = list(client.query(probe, job_config=no_cache).result())[0]['dims']
        streak += 1
        print(f'  probe {streak}/{required_streak} succeeded ({waited}s elapsed): {list(dims)}')
    except BadRequest as e:
        if 'requires reservation' not in str(e):
            raise
        streak = 0
        print(f'  not propagated yet ({waited}s elapsed)')
    if streak < required_streak:
        if waited >= 600:
            raise RuntimeError('Reservation did not become usable within 10 minutes')
        time.sleep(15)
        waited += 15

print('Reservation is live.')

Created projects/statmike-mlops-349915/locations/US/reservations/bqml-image-temp


Created projects/statmike-mlops-349915/locations/US/reservations/bqml-image-temp/assignments/11619011449427980529
  not propagated yet (0s elapsed)


  not propagated yet (15s elapsed)


  not propagated yet (30s elapsed)


  not propagated yet (45s elapsed)


  not propagated yet (60s elapsed)


  not propagated yet (75s elapsed)


  probe 1/3 succeeded (90s elapsed): [1, 2, 3]


  probe 2/3 succeeded (105s elapsed): [1, 2, 3]


  probe 3/3 succeeded (120s elapsed): [1, 2, 3]
Reservation is live.



---
## Step 1 — What `ML.DECODE_IMAGE` actually returns

`[1, 2, 3]` from the cell above is already the first correction. The reference pages describe the output as `STRUCT<ARRAY<INT64> shape, ARRAY<FLOAT64> values>`. Select `.shape` and the query fails; the field is named **`dimensions`**.

Note also what the call above took as input: a `BYTES` literal built by `FROM_BASE64`. Every doc example passes an object table's `data` pseudocolumn, and the parameter is described as coming from one, but the signature is just `BYTES` — so the function is testable without a bucket, a connection, or an object table. Step 8 uses the object-table path for real; everything between here and there uses literals, which is why every value below is reproducible without any Cloud Storage setup.

In [8]:
query = f"""
SELECT {img('swatch.png')}.dimensions AS dimensions,
       {img('swatch.png')}.values     AS values
"""
row = list(client.query(query, job_config=no_cache).result())[0]
h, w, c = row['dimensions']
print(f'dimensions = {list(row["dimensions"])}  ->  height={h}, width={w}, channels={c}')
print(f'values     = {list(row["values"])}')
print(f'len(values) = {len(row["values"])} = {h} x {w} x {c}')

# Ground truth: the array this notebook built, flattened row-major, divided by 255.
expected = swatch_a.astype(np.float64).ravel() / 255.0
print(f'\nexpected   = {list(expected)}')
print(f'max |BigQuery - expected| = {np.abs(np.array(row["values"]) - expected).max()}')

dimensions = [1, 2, 3]  ->  height=1, width=2, channels=3
values     = [0.7843137254901961, 0.39215686274509803, 0.19607843137254902, 0.0, 0.5019607843137255, 1.0]
len(values) = 6 = 1 x 2 x 3

expected   = [0.7843137254901961, 0.39215686274509803, 0.19607843137254902, 0.0, 0.5019607843137255, 1.0]
max |BigQuery - expected| = 0.0


Three parts of the contract, all confirmed:

- **`dimensions` is `[height, width, channels]`** — the 1×2 image gives `[1, 2, 3]`, so height leads.
- **`values` is flattened row-major in HWC order** — pixel 1's three channels, then pixel 2's. Not planar (all R, then all G, then all B), which is the other convention a vision model might want and which you would have to build yourself.
- **The scaling is exactly `pixel / 255`, with no rounding error at all.** The difference from the float64 ground truth is `0.0`, not a small epsilon. `200/255` came back as the correctly-rounded double.

That last point matters more than it looks: it means these values are exact enough to reason about, and it sets up the precision comparison in Step 6, where one of the four functions behaves differently.


---
## Step 2 — Decode always returns exactly 3 channels

The reference pages say the input may be JPEG, PNG, or BMP. They do not say what happens to the channel count when the source image is not 3-channel RGB. Two cases matter in practice — a PNG with an alpha channel, and a single-channel grayscale PNG — and they resolve in opposite directions.

In [9]:
query = f"""
SELECT {img('alpha.png')}.dimensions AS rgba_dimensions,
       {img('alpha.png')}.values     AS rgba_values,
       {img('gray.png')}.dimensions  AS gray_dimensions,
       {img('gray.png')}.values      AS gray_values
"""
row = list(client.query(query, job_config=no_cache).result())[0]

print(f'Source alpha.png is RGBA, shape {alpha_a.shape}, pixels {alpha_a.reshape(-1, 4).tolist()}')
print(f'  decoded dimensions: {list(row["rgba_dimensions"])}')
print(f'  decoded x255:       {[round(v * 255, 6) for v in row["rgba_values"]]}')

gray_source = np.array(Image.fromarray(swatch_a).convert('L'))
print(f'\nSource gray.png is single-channel L, shape {gray_source.shape}, pixels {gray_source.ravel().tolist()}')
print(f'  decoded dimensions: {list(row["gray_dimensions"])}')
print(f'  decoded x255:       {[round(v * 255, 6) for v in row["gray_values"]]}')

Source alpha.png is RGBA, shape (1, 2, 4), pixels [[200, 100, 50, 255], [0, 128, 255, 128]]
  decoded dimensions: [1, 2, 3]
  decoded x255:       [200.0, 100.0, 50.0, 0.0, 128.0, 255.0]

Source gray.png is single-channel L, shape (1, 2), pixels [124, 104]
  decoded dimensions: [1, 2, 3]
  decoded x255:       [124.0, 124.0, 124.0, 104.0, 104.0, 104.0]


**Channels is always 3.** Both images decode to `channels = 3` regardless of what they contained.

- **The alpha channel is dropped, not applied.** The second pixel of `alpha.png` has alpha 128 — half transparent — and its RGB values come back as `0, 128, 255`, identical to the fully-opaque source values. Nothing was blended against a background and nothing was premultiplied. If your corpus mixes transparent PNGs with opaque ones, BigQuery will hand your model the raw RGB underneath the transparency, silently, and the two will not be comparable to whatever your training-time preprocessing did if that preprocessing composited. There is no flag to change this and no warning.
- **A grayscale image is replicated into three identical channels.** `gray.png` holds one value per pixel and decodes to that value three times. The size of a decoded feature vector therefore depends only on height and width, never on the source encoding — convenient for a fixed-shape model, and it means a mixed corpus of RGB, RGBA, and grayscale files needs no per-file branching. (This is a rule about *decoding*. Step 5 shows `ML.CONVERT_COLOR_SPACE(..., 'GRAYSCALE')` going the other way, back down to 1 channel.)

The format list is enforced at decode time, and the error names the accepted set:

In [10]:
try:
    client.query("SELECT ML.DECODE_IMAGE(b'not an image at all').dimensions", job_config=no_cache).result()
except BadRequest as e:
    print(str(e).split(';')[0])

400 GET https://bigquery.googleapis.com/bigquery/v2/projects/statmike-mlops-349915/queries/08f8454e-fea8-4e23-a2e2-7cb18f1c244f?maxResults=0&location=US&prettyPrint=false: Input image bytes could not be decoded. Expected valid image of type PNG, JPEG, or BMP


PNG, JPEG, BMP — no GIF, no WEBP, no TIFF, no AVIF. An object table happily lists files of any type; the failure surfaces only when a row reaches `ML.DECODE_IMAGE`, which means one stray file in a bucket fails the whole query rather than that one row.

**JPEG decodes to different numbers than PNG for the same source image**, which is worth stating explicitly because it is easy to build a corpus in mixed formats and treat the results as comparable:

In [11]:
query = f"""
SELECT ML.CONVERT_IMAGE_TYPE({img('swatch.png')}).values AS png_ints,
       ML.CONVERT_IMAGE_TYPE({img('swatch.jpg')}).values AS jpg_ints
"""
row = list(client.query(query, job_config=no_cache).result())[0]
print(f'source pixels : {swatch_a.reshape(-1, 3).tolist()}')
print(f'via PNG       : {np.array(row["png_ints"]).reshape(-1, 3).tolist()}')
print(f'via JPEG      : {np.array(row["jpg_ints"]).reshape(-1, 3).tolist()}')

source pixels : [[200, 100, 50], [0, 128, 255]]
via PNG       : [[200, 100, 50], [0, 128, 255]]
via JPEG      : [[105, 123, 159], [86, 104, 140]]


PNG is lossless and round-trips the source exactly. JPEG does not, and on a 2-pixel image it is not close — JPEG's 8×8 DCT blocks and chroma subsampling have nothing to work with at this size, so the reconstruction is dominated by block averaging. A photograph would survive far better. The point stands either way: **the decoded feature vector is a property of the file, not of the picture**, so re-encoding a corpus changes your features.


---
## Step 3 — The documented value ranges are wrong at the top

Every reference page states the decoded range as `[0, 1)` and the `ML.CONVERT_IMAGE_TYPE` range as `[0, 255)`. Both are written half-open, excluding the upper bound. A pure-white pixel settles it.

In [12]:
query = f"""
SELECT {img('white.png')}.values                             AS float_values,
       ML.CONVERT_IMAGE_TYPE({img('white.png')}).values      AS int_values,
       {img('white.png')}.values[OFFSET(0)] = 1.0            AS float_is_exactly_one,
       ML.CONVERT_IMAGE_TYPE({img('white.png')}).values[OFFSET(0)] = 255 AS int_is_exactly_255
"""
row = list(client.query(query, job_config=no_cache).result())[0]
for k, v in row.items():
    print(f'{k:22s} {list(v) if isinstance(v, list) else v}')

float_values           [1.0, 1.0, 1.0]
int_values             [255, 255, 255]
float_is_exactly_one   True
int_is_exactly_255     True


Both endpoints are attainable. The decoded range is **`[0, 1]`** and the integer range is **`[0, 255]`**, both closed. The comparison is done in SQL rather than by eyeballing a printed `1.0`, because a float that prints as `1.0` need not equal `1.0`; this one does.

This is not pedantry. `[0, 255)` is how you would describe a range that a model's input validation should reject 255 from, and code written to that spec — a clamp, an assertion, a bucket boundary — is wrong on every white pixel, which is a common pixel.


---
## Step 4 — `ML.CONVERT_IMAGE_TYPE` is exact, and one-way

Its job is the float→integer direction for models that want `uint8` input. Round-tripping the swatch shows it is lossless in that direction:

In [13]:
query = f'SELECT ML.CONVERT_IMAGE_TYPE({img("swatch.png")}).values AS ints'
ints = list(list(client.query(query, job_config=no_cache).result())[0]['ints'])
print(f'source : {swatch_a.ravel().tolist()}')
print(f'decoded -> converted back to int : {ints}')
print(f'identical: {ints == swatch_a.ravel().tolist()}')

source : [200, 100, 50, 0, 128, 255]
decoded -> converted back to int : [200, 100, 50, 0, 128, 255]
identical: True


`pixel / 255` then back is exact for all 256 byte values, so decode → convert is a safe round trip.

The reverse does not exist. Feeding integer image data back in is a signature error, not a runtime one — and the error text is the authoritative statement of the struct's real field names:

In [14]:
query = f'SELECT ML.CONVERT_IMAGE_TYPE(ML.CONVERT_IMAGE_TYPE({img("swatch.png")})).values'
try:
    client.query(query, job_config=no_cache).result()
except BadRequest as e:
    print(str(e).split(';')[0])

400 No matching signature for function BIGQUERYGOOGLESQL:ML.CONVERT_IMAGE_TYPE
  Argument types: STRUCT<dimensions ARRAY<INT64>, values ARRAY<INT64>>
  Signature: ML.CONVERT_IMAGE_TYPE(STRUCT<dimensions ARRAY<INT64>, values ARRAY<FLOAT64>>)
    Argument 1: Unable to coerce type STRUCT<dimensions ARRAY<INT64>, values ARRAY<INT64>> to expected type STRUCT<dimensions ARRAY<INT64>, values ARRAY<FLOAT64>> at [1:8]


`STRUCT<dimensions ARRAY<INT64>, values ARRAY<FLOAT64>>` in, `STRUCT<dimensions ARRAY<INT64>, values ARRAY<INT64>>` out. `dimensions`, from the engine itself.

The practical consequence: **`ML.CONVERT_IMAGE_TYPE` is terminal.** Once converted, the only remaining function you can apply is `ML.RESIZE_IMAGE` (Step 6 shows it accepts either value type) — `ML.CONVERT_COLOR_SPACE` also accepts integers, but nothing can put you back in the float domain. Order your pipeline accordingly: decode → resize → color space → convert type, last.


---
## Step 5 — `ML.CONVERT_COLOR_SPACE`, and the two different lumas

Four targets: `GRAYSCALE`, `HSV`, `YIQ`, `YUV`. The interesting one is that two of them compute a luminance channel, and they do not agree.

In [15]:
spaces = ['GRAYSCALE', 'HSV', 'YIQ', 'YUV']
select = ',\n  '.join(
    f"ML.CONVERT_COLOR_SPACE({img('swatch.png')}, '{s}').dimensions AS {s.lower()}_dims,\n  "
    f"ML.CONVERT_COLOR_SPACE({img('swatch.png')}, '{s}').values     AS {s.lower()}"
    for s in spaces)
query = f'SELECT\n  {select}'
row = list(client.query(query, job_config=no_cache).result())[0]

dims = {s: list(row[f'{s.lower()}_dims']) for s in spaces}
converted = {s: np.array(row[s.lower()]).reshape(dims[s]) for s in spaces}
for s in spaces:
    print(f'{s:10s} dimensions={dims[s]}')
    for i, pixel in enumerate(converted[s].reshape(-1, dims[s][-1])):
        print(f'{"":12s}pixel {i} (RGB {swatch_a.reshape(-1, 3)[i].tolist()}): {pixel.tolist()}')

GRAYSCALE  dimensions=[1, 2, 1]
            pixel 0 (RGB [200, 100, 50]): [0.48698039215686273]
            pixel 1 (RGB [0, 128, 255]): [0.40865098039215686]
HSV        dimensions=[1, 2, 3]
            pixel 0 (RGB [200, 100, 50]): [0.05555555555555555, 0.75, 0.7843137254901961]
            pixel 1 (RGB [0, 128, 255]): [0.5830065359477123, 1.0, 1.0]
YIQ        dimensions=[1, 2, 3]
            pixel 0 (RGB [200, 100, 50]): [0.4870588235294117, 0.2966951176470588, 0.021907296078431334]
            pixel 1 (RGB [0, 128, 255]): [0.40865098039215686, -0.45916060141176473, 0.04880649211764704]
YUV        dimensions=[1, 2, 3]
            pixel 0 (RGB [200, 100, 50]): [0.4870588235294117, -0.14319465294117648, 0.2607766705882352]
            pixel 1 (RGB [0, 128, 255]): [0.40865098039215686, 0.29100935988235294, -0.3585025555294118]


**`GRAYSCALE` changes the channel count — it returns 1, not 3.** That is the opposite of `ML.DECODE_IMAGE`, which in Step 2 replicated a single-channel grayscale PNG *up* to 3. So a decode-then-grayscale pipeline goes 1 → 3 → 1, and the "always 3 channels" rule from Step 2 is a rule about *decoding*, not about the family. `HSV`, `YIQ`, and `YUV` keep 3 channels and fill all three with genuinely distinct values.

`HSV`'s hue is on **`[0, 1]`, not degrees**: pixel 1's hue of ~0.0556 is 20°/360°, and pixel 2's ~0.583 is 210°/360°. A model or a threshold expecting degrees is off by a factor of 360.

Now the part worth stopping on. `GRAYSCALE`'s value and `YIQ`'s first channel are both luma, both computed from the same RGB, and they are different numbers. Which weights does each use?

In [16]:
bt601   = np.array([0.299, 0.587, 0.114])       # the true ITU-R BT.601 luma coefficients
rounded = np.array([0.2989, 0.5870, 0.1140])     # the 4-decimal variant TensorFlow's rgb_to_grayscale uses

rgb = swatch_a.astype(np.float64).reshape(-1, 3) / 255.0
bq_grayscale = converted['GRAYSCALE'][..., 0].ravel()
bq_yiq_y     = converted['YIQ'][..., 0].ravel()

rows = []
for label, actual in [('GRAYSCALE', bq_grayscale), ('YIQ channel Y', bq_yiq_y)]:
    rows.append({
        'BigQuery output': label,
        'max diff vs [0.299, 0.587, 0.114]':    np.abs(actual - rgb @ bt601).max(),
        'max diff vs [0.2989, 0.5870, 0.1140]': np.abs(actual - rgb @ rounded).max(),
    })
display(pd.DataFrame(rows))

gap = bq_grayscale - bq_yiq_y
print(f'GRAYSCALE - YIQ.Y per pixel: {gap.tolist()}')
print(f'as 8-bit levels (x255):      {(gap * 255).tolist()}')

,BigQuery output,"max diff vs [0.299, 0.587, 0.114]","max diff vs [0.2989, 0.5870, 0.1140]"
0,GRAYSCALE,0.000078,5.551115e-17
1,YIQ channel Y,0.000000,7.843137e-05


GRAYSCALE - YIQ.Y per pixel: [-7.843137254898158e-05, 0.0]
as 8-bit levels (x255):      [-0.019999999999990303, 0.0]


Two different lumas, each matching one weight vector to machine epsilon and neither matching the other:

- **`GRAYSCALE` uses `[0.2989, 0.5870, 0.1140]`** — the 4-decimal rounding that TensorFlow's `tf.image.rgb_to_grayscale` hardcodes.
- **`YIQ`'s Y uses `[0.299, 0.587, 0.114]`** — the actual BT.601 coefficients, exactly.

The gap is real but tiny — a fraction of one 8-bit level. It will never be visible and it will never change a classification. It matters for exactly one thing, which is worth knowing before it costs you an afternoon: **`GRAYSCALE` and `ML.CONVERT_COLOR_SPACE(..., 'YIQ')`'s Y are not interchangeable in a bit-for-bit reproducibility check.** If you are diffing a BigQuery preprocessing pipeline against a reimplementation elsewhere, you have to match the right one.

Integer-valued image data is accepted too, and is normalized on the way in rather than treated as already-scaled:

In [17]:
query = f"""
SELECT ML.CONVERT_COLOR_SPACE({img('swatch.png')}, 'HSV').values                        AS from_float,
       ML.CONVERT_COLOR_SPACE(ML.CONVERT_IMAGE_TYPE({img('swatch.png')}), 'HSV').values AS from_int
"""
row = list(client.query(query, job_config=no_cache).result())[0]
print(f'from FLOAT64 input : {list(row["from_float"])}')
print(f'from INT64 input   : {list(row["from_int"])}')
print(f'identical: {list(row["from_float"]) == list(row["from_int"])}')

from FLOAT64 input : [0.05555555555555555, 0.75, 0.7843137254901961, 0.5830065359477123, 1.0, 1.0]
from INT64 input   : [0.05555555555555555, 0.75, 0.7843137254901961, 0.5830065359477123, 1.0, 1.0]
identical: True


Identical, and the output is `FLOAT64` either way. So `ML.CONVERT_COLOR_SPACE` divides integer input by 255 before converting — it never mistakes a `200` for a value already on `[0, 1]`. That is the safe behavior, and it means the pipeline order from Step 4 is a recommendation rather than a correctness requirement for this function.

An unrecognized target is rejected by name:

In [18]:
try:
    client.query(f"SELECT ML.CONVERT_COLOR_SPACE({img('swatch.png')}, 'LAB').values",
                 job_config=no_cache).result()
except BadRequest as e:
    print(str(e).split(';')[0])

400 Color space LAB is not supported



---
## Step 6 — `ML.RESIZE_IMAGE`: bilinear, float32, and type-preserving

The 4×8 gradient is the test image here — big enough to have interior pixels, small enough to print. Start with what `preserve_aspect_ratio` actually does to the output dimensions.

In [19]:
targets = [(4, 4, True), (3, 3, True), (2, 2, True), (9, 9, True), (4, 4, False)]
select = ',\n  '.join(
    f'ML.RESIZE_IMAGE({img("gradient.png")}, {h}, {w}, {str(p).upper()}).dimensions AS t{i}'
    for i, (h, w, p) in enumerate(targets))
row = list(client.query(f'SELECT\n  {select}', job_config=no_cache).result())[0]

print(f'source dimensions: {list(gradient_a.shape)}\n')
out = []
for i, (h, w, p) in enumerate(targets):
    dims = list(row[f't{i}'])
    scale = min(h / gradient_a.shape[0], w / gradient_a.shape[1]) if p else None
    out.append({'target (h, w)': f'{h} x {w}', 'preserve_aspect_ratio': p,
                'implied scale': round(scale, 4) if scale else '—',
                'unrounded h x w': f'{gradient_a.shape[0] * scale:g} x {gradient_a.shape[1] * scale:g}' if scale else '—',
                'result dimensions': dims})
display(pd.DataFrame(out))

source dimensions: [4, 8, 3]



,"target (h, w)",preserve_aspect_ratio,implied scale,unrounded h x w,result dimensions
0,4 x 4,True,0.5,2 x 4,"[2, 4, 3]"
1,3 x 3,True,0.375,1.5 x 3,"[2, 3, 3]"
2,2 x 2,True,0.25,1 x 2,"[1, 2, 3]"
3,9 x 9,True,1.125,4.5 x 9,"[4, 9, 3]"
4,4 x 4,False,—,—,"[4, 4, 3]"


With `preserve_aspect_ratio = TRUE` the two target values are a **bounding box, not a size**: the scale is `min(target_h / h, target_w / w)` and the result is the largest image inside the box at the original aspect ratio. With `FALSE` you get exactly what you asked for, distortion included.

Look at the two rows where the scaled dimension lands on a half. `4 × 8` into a `3 × 3` box scales by 0.375, giving an unrounded height of 1.5, and the result is **2**. Into a `9 × 9` box it scales by 1.125, giving an unrounded height of 4.5, and the result is **4**. Half rounds up in one case and down in the other — that is round-half-to-even, the rounding NumPy and TensorFlow use and not the one SQL's `ROUND()` uses. Step 7 shows why that is exactly what you would expect.

Now the pixel values. A 4×8 down to 2×4 is a clean 2:1 in both directions:

In [20]:
query = f'SELECT ML.RESIZE_IMAGE({img("gradient.png")}, 2, 4, FALSE).values AS v'
resized = np.array(list(client.query(query, job_config=no_cache).result())[0]['v']).reshape(2, 4, 3)
print('red channel of the resized 2x4 (x255):')
print(np.round(resized[..., 0] * 255, 6))
print('\nred channel of the 4x8 source (x255):')
print(gradient_a[..., 0])
print(f'\nsingle value, full precision: {resized[0, 0, 0]!r}')
print(f'nearest float32:              {np.float32(resized[0, 0, 0]).item()!r}')
print(f'exactly representable in float32: {np.float64(np.float32(resized[0, 0, 0])) == resized[0, 0, 0]}')

red channel of the resized 2x4 (x255):
[[ 16.000001  80.000003 144.000007 208.000003]
 [ 16.000001  80.000003 144.000007 208.000003]]

red channel of the 4x8 source (x255):
[[  0  32  64  96 128 160 192 224]
 [  0  32  64  96 128 160 192 224]
 [  0  32  64  96 128 160 192 224]
 [  0  32  64  96 128 160 192 224]]

single value, full precision: 0.062745101749897
nearest float32:              0.062745101749897
exactly representable in float32: True


Two things.

**The interpolation is bilinear with half-pixel centers.** Output pixel `[0,0]` sits between source columns 0 and 1 (red values 0 and 32) and between source rows 0 and 1, and comes back at 16/255 — the average of the four neighbors, not a copy of the top-left pixel. That is `align_corners = False` sampling, the modern default.

**The values are float32, unlike everything else in this family.** Every number `ML.DECODE_IMAGE` and `ML.CONVERT_COLOR_SPACE` returned above was float64-exact; this one is a float64 that exactly equals its float32 neighbor, which is the signature of a value computed in single precision and widened on the way out. It costs about seven significant digits. Irrelevant for feeding a model — models are float32 anyway — and decisive if you are diffing against a float64 reimplementation, which is the trap Step 5's luma finding also sets.

Resize is the one function in the family that **preserves the value type** rather than forcing float:

In [21]:
query = f"""
SELECT ML.RESIZE_IMAGE(ML.CONVERT_IMAGE_TYPE({img('swatch.png')}), 1, 1, FALSE).values AS int_in_int_out,
       ML.RESIZE_IMAGE({img('swatch.png')}, 1, 1, FALSE).values                        AS float_in_float_out
"""
row = list(client.query(query, job_config=no_cache).result())[0]
print(f'source pixels        : {swatch_a.reshape(-1, 3).tolist()}')
print(f'exact channel means  : {swatch_a.astype(np.float64).reshape(-1, 3).mean(axis=0).tolist()}')
print(f'INT64 in -> out      : {list(row["int_in_int_out"])}')
print(f'FLOAT64 in -> out x255: {[v * 255 for v in row["float_in_float_out"]]}')

source pixels        : [[200, 100, 50], [0, 128, 255]]
exact channel means  : [100.0, 114.0, 152.5]
INT64 in -> out      : [100, 114, 152]
FLOAT64 in -> out x255: [100.00000163912773, 114.00000840425491, 152.49999850988388]


Integers in, integers out. The blue channel's true mean is 152.5 and it comes back as 152 — half-to-even again, consistent with the dimension rounding above. Feeding integers to resize therefore quantizes at every resize step; keep the pipeline in floats and convert last, as Step 4 concluded for a different reason.

### The failure mode worth knowing about

Ask for an aspect-preserving fit whose scale rounds a dimension down to zero — the 4×8 gradient into a 1×1 box scales by 0.125, giving a height of 0.5 — and the query fails.

> **The client library will retry this for minutes if you let it.** BigQuery reports it as a generic internal error with the standard "this is usually caused by a transient issue" advice, and `google-cloud-bigquery`'s default job-retry policy believes it: it resubmits the job, which fails identically, repeatedly. It is fully deterministic and no amount of retrying will change it. **`job_retry=None` is what makes this cell return in seconds instead of hanging.**

> **Catch the base class, and read the failure off the job.** The same deterministic failure has surfaced here as both `InternalServerError` (HTTP 500) and `BadRequest` (HTTP 400) on different runs, depending on which API call observes the dead job, and the 400 form arrives with a generic message that omits the diagnostic code. `job.error_result['message']` carries the full text either way, so catch `GoogleAPICallError` — the parent of both — and print the job's own error rather than the exception's.

In [22]:
from google.api_core.exceptions import GoogleAPICallError

query = f'SELECT ML.RESIZE_IMAGE({img("gradient.png")}, 1, 1, TRUE).dimensions'
started = time.time()
job = client.query(query, job_config=no_cache, job_retry=None)
try:
    job.result()
    print('Unexpectedly succeeded')
except GoogleAPICallError as e:
    print(f'Failed in {time.time() - started:.1f}s as {type(e).__name__} (HTTP {e.code}).\n')
    detail = job.error_result
    if not detail:
        job.reload()
        detail = job.error_result
    print((detail or {}).get('message', str(e)))

Failed in 4.3s as BadRequest (HTTP 400).

An internal error occurred and the request could not be completed. This is usually caused by a transient issue. Retrying the job with back-off as described in the BigQuery SLA should solve the problem: https://cloud.google.com/bigquery/sla. If the error continues to occur please contact support at https://cloud.google.com/support. Error: 80038528


The reproducible input rule is: **`preserve_aspect_ratio = TRUE` fails whenever `min(target_h / h, target_w / w)` scales either source dimension to less than 0.5**, i.e. whenever it rounds to zero. Verified on two differently-shaped sources; the `2 × 2` fit two cells up is the same target size on the same image with a non-degenerate result, so it is the rounding-to-zero and not the small target that does it.

What produces the internal error rather than a clean argument error is not observable from outside. A zero-sized dimension reaching an op that cannot accept one would explain it, and so would several other things; the message says nothing either way. What is worth carrying forward is operational, not diagnostic: **a "transient" internal error from these functions may be perfectly deterministic**, and the guard is to bound the target box so no dimension can round to zero — check `min(th/h, tw/w) * min(h, w) >= 0.5` before the query, or simply pass `FALSE` and accept the distortion.


---
## Step 7 — All four are TensorFlow image ops

Every finding above has the same shape: the behavior is not what BigQuery's documentation says, and it is exactly what TensorFlow does. Rather than infer that from a family resemblance, diff it directly — run the conversions in BigQuery, run the corresponding `tf.image` op locally, and subtract.

In [23]:
import tensorflow as tf

query = f"""
SELECT
  ML.CONVERT_COLOR_SPACE({img('swatch.png')}, 'GRAYSCALE').values AS grayscale,
  ML.CONVERT_COLOR_SPACE({img('swatch.png')}, 'HSV').values       AS hsv,
  ML.CONVERT_COLOR_SPACE({img('swatch.png')}, 'YIQ').values       AS yiq,
  ML.CONVERT_COLOR_SPACE({img('swatch.png')}, 'YUV').values       AS yuv,
  ML.RESIZE_IMAGE({img('gradient.png')}, 2, 4, FALSE).values      AS resize_even,
  ML.RESIZE_IMAGE({img('gradient.png')}, 3, 5, FALSE).values      AS resize_odd,
  ML.CONVERT_IMAGE_TYPE({img('swatch.png')}).values               AS to_uint8
"""
row = list(client.query(query, job_config=no_cache).result())[0]

swatch_f64 = swatch_a.astype(np.float64) / 255.0
grad_f64 = gradient_a.astype(np.float64) / 255.0

# Pin the tensor dtype explicitly. BigQuery computes the color-space conversions in
# double, so the reference has to as well -- and handing tf.image a bare NumPy float64
# array does not reliably get you there (the `reference dtype` column below is the check).
swatch_t64 = tf.constant(swatch_f64, tf.float64)

cases = [
    ('ML.CONVERT_COLOR_SPACE GRAYSCALE', row['grayscale'],
        tf.image.rgb_to_grayscale(swatch_t64).numpy(),
        'tf.image.rgb_to_grayscale, called as-is'),
    ('ML.CONVERT_COLOR_SPACE GRAYSCALE', row['grayscale'],
        (swatch_f64 @ rounded)[..., None],
        'the same weights, evaluated in float64'),
    ('ML.CONVERT_COLOR_SPACE HSV', row['hsv'],
        tf.image.rgb_to_hsv(swatch_t64).numpy(), 'tf.image.rgb_to_hsv (float64)'),
    ('ML.CONVERT_COLOR_SPACE YIQ', row['yiq'],
        tf.image.rgb_to_yiq(swatch_t64).numpy(), 'tf.image.rgb_to_yiq (float64)'),
    ('ML.CONVERT_COLOR_SPACE YUV', row['yuv'],
        tf.image.rgb_to_yuv(swatch_t64).numpy(), 'tf.image.rgb_to_yuv (float64)'),
    # Resize runs in float32 (Step 6), so cast the input to match.
    ('ML.RESIZE_IMAGE 4x8 -> 2x4', row['resize_even'],
        tf.image.resize(tf.constant(grad_f64, tf.float32), [2, 4], method='bilinear').numpy(),
        'tf.image.resize bilinear (float32)'),
    ('ML.RESIZE_IMAGE 4x8 -> 3x5', row['resize_odd'],
        tf.image.resize(tf.constant(grad_f64, tf.float32), [3, 5], method='bilinear').numpy(),
        'tf.image.resize bilinear (float32)'),
    ('ML.CONVERT_IMAGE_TYPE', row['to_uint8'],
        tf.image.convert_image_dtype(swatch_t64, tf.uint8).numpy(),
        'tf.image.convert_image_dtype -> uint8'),
]

EPS64, EPS32 = np.finfo(np.float64).eps, np.finfo(np.float32).eps

def compare(bq, ref):
    """Diff, plus the precision the gap is consistent with."""
    bq, ref = np.array(bq, dtype=np.float64), ref.astype(np.float64).ravel()
    diff, scale = float(np.abs(bq - ref).max()), float(np.abs(bq).max())
    if diff == 0.0:
        return diff, 'exact'
    if diff <= 4 * EPS64 * scale:
        return diff, 'double precision (<= a few ULP)'
    if diff <= 4 * EPS32 * scale:
        return diff, 'single precision (float32 somewhere in the op)'
    return diff, 'a different answer'

table = []
for name, bq, ref, op in cases:
    diff, verdict = compare(bq, ref)
    table.append({'BigQuery function': name, 'TensorFlow op': op,
                  'reference dtype': str(ref.dtype), 'max |difference|': diff,
                  'consistent with': verdict})
display(pd.DataFrame(table))

single = [t['BigQuery function'] for t in table if t['consistent with'].startswith('single')]
print(f'TensorFlow {tf.__version__}')
print(f'single-precision rows: {", ".join(single) if single else "none"}')

# Which side is the single precision on? Compare BigQuery to BigQuery: the HSV values
# from this multi-function query against the ones Step 5 got with HSV alone.
delta = float(np.abs(np.array(row['hsv']) - np.array(converted['HSV']).ravel()).max())
print(f'BigQuery HSV here vs Step 5 (HSV alone in its query): max |difference| {delta:.6e}')

I0000 00:00:1789147565.550971 1698405 cudart_stub.cc:31] Could not find cuda drivers on your machine, GPU will not be used.
I0000 00:00:1789147565.628595 1698405 cpu_feature_guard.cc:227] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.


I0000 00:00:1789147568.791362 1698405 cudart_stub.cc:31] Could not find cuda drivers on your machine, GPU will not be used.


E0000 00:00:1789147574.297020 1698405 cuda_platform.cc:52] failed call to cuInit: INTERNAL: CUDA error: Failed call to cuInit: UNKNOWN ERROR (303)


,BigQuery function,TensorFlow op,reference dtype,max |difference|,consistent with
0,ML.CONVERT_COLOR_SPACE GRAYSCALE,"tf.image.rgb_to_grayscale, called as-is",float64,1.627324e-08,single precision (float32 somewhere in the op)
1,ML.CONVERT_COLOR_SPACE GRAYSCALE,"the same weights, evaluated in float64",float64,5.551115e-17,double precision (<= a few ULP)
2,ML.CONVERT_COLOR_SPACE HSV,tf.image.rgb_to_hsv (float64),float64,0.000000e+00,exact
3,ML.CONVERT_COLOR_SPACE YIQ,tf.image.rgb_to_yiq (float64),float64,5.551115e-17,double precision (<= a few ULP)
4,ML.CONVERT_COLOR_SPACE YUV,tf.image.rgb_to_yuv (float64),float64,5.551115e-17,double precision (<= a few ULP)
5,ML.RESIZE_IMAGE 4x8 -> 2x4,tf.image.resize bilinear (float32),float32,0.000000e+00,exact
6,ML.RESIZE_IMAGE 4x8 -> 3x5,tf.image.resize bilinear (float32),float32,0.000000e+00,exact
7,ML.CONVERT_IMAGE_TYPE,tf.image.convert_image_dtype -> uint8,uint8,0.000000e+00,exact


TensorFlow 2.21.0
single-precision rows: ML.CONVERT_COLOR_SPACE GRAYSCALE
BigQuery HSV here vs Step 5 (HSV alone in its query): max |difference| 0.000000e+00


**`ML.CONVERT_COLOR_SPACE(..., 'HSV')` and both `ML.RESIZE_IMAGE` cases are bit-exact** — a difference of literally zero, including the 4×8 → 3×5 resize where the scale factor is not an integer ratio in either direction. `ML.CONVERT_IMAGE_TYPE` is exact too, since it produces integers.

`YIQ` and `YUV` land a last-place unit of a double away from `tf.image`'s answer — the residue of doing the same arithmetic in a different order, not a different formula.

`GRAYSCALE` is the row with a documented story. Called the ordinary way, `tf.image.rgb_to_grayscale` is eight orders of magnitude further off than `YIQ` is, and the reason is in TensorFlow's own source: it does `flt_image = convert_image_dtype(images, dtypes.float32)` before applying the weights, then converts back — so the returned tensor is still a double, but the arithmetic ran in **float32**. Apply the identical weights in float64 and the gap collapses to the same last-place residue as everything else. **BigQuery runs this in double where TensorFlow does not** — the algorithm is TensorFlow's (Step 5 pinned the weights to its 4-decimal constants), the precision is better than TensorFlow's.

That is also why the `consistent with` column is worth more than the raw magnitude: a ~1e-8 gap is not "a different formula," it is the resolution of a single-precision float, and it says only that *someone* on one side of the comparison spent a float32.

**The `reference dtype` column is there because this comparison is easy to get wrong, and the wrong version looks like a finding.** Hand `tf.image` a bare NumPy `float64` array rather than an explicitly typed tensor and the op can return **float32** — which turns the bit-exact `HSV` row into one the table classifies as single precision, at the same magnitude as the real `GRAYSCALE` result above it — an artifact of the measurement setup, indistinguishable from a property of the engine. That is why the cell pins its reference tensor with `tf.constant(..., tf.float64)` and prints the dtype it actually got. The last line of the cell is the control on the other side: BigQuery returns the **identical** `HSV` values whether the conversion sits alone in a query or beside six other image functions, so the precision of these functions is a property of the function, not of the query it lands in.

This is the practical takeaway of the whole notebook. When the reference page is silent or wrong — the channel count, the closed ranges, the rounding rule, the luma weights, the interpolation convention — **the answer is what `tf.image` does**, and `tf.image`'s documentation is specific where BigQuery's is not.


---
## Step 8 — The real path: an object table over Cloud Storage

Everything above used `BYTES` literals to keep the values reproducible. Production input is an [object table](https://cloud.google.com/bigquery/docs/object-tables): a read-only table over files in Cloud Storage whose `data` pseudocolumn is the file's bytes.

Unlike the functions themselves, an object table **does** need a Cloud Resource Connection, and the connection's service account needs read access to the bucket.

> **This changes IAM on your project** — it grants `roles/storage.objectViewer` to one newly-created, narrowly-scoped service account. The grant is checked explicitly rather than assumed: a silently-swallowed failure here surfaces as a confusing permission error on the query two cells later.

> **The bucket is never created or deleted by this notebook** — set `BUCKET` above to one you already own in the same location as `DATASET_ID`. Cleanup removes only the objects this notebook wrote, under the `bq_ml/image/` prefix.

In [24]:
import json as _json

prefix = 'bq_ml/image/'
for name, data in raw.items():
    blob = gcs_bucket.blob(prefix + name)
    blob.upload_from_string(data, content_type='image/jpeg' if name.endswith('.jpg') else 'image/png')
print(f'Uploaded {len(raw)} objects to gs://{BUCKET}/{prefix}')

r_conn = subprocess.run(
    ['bq', 'mk', '--connection', '--location', LOCATION, '--connection_type', 'CLOUD_RESOURCE',
     '--project_id', PROJECT_ID, CONNECTION_ID],
    capture_output=True, text=True,
)
if r_conn.returncode != 0 and 'already exists' not in r_conn.stderr.lower():
    raise RuntimeError(f'Failed to create connection {CONNECTION_ID}:\n{r_conn.stderr}')

r_show = subprocess.run(
    ['bq', 'show', '--connection', '--format=json', '--project_id', PROJECT_ID, '--location', LOCATION, CONNECTION_ID],
    capture_output=True, text=True, check=True,
)
connection_sa = _json.loads(r_show.stdout)['cloudResource']['serviceAccountId']
print('Connection service account:', connection_sa)

r_iam = subprocess.run(
    ['gcloud', 'projects', 'add-iam-policy-binding', PROJECT_ID,
     f'--member=serviceAccount:{connection_sa}', '--role=roles/storage.objectViewer', '--quiet'],
    capture_output=True, text=True,
)
if r_iam.returncode != 0:
    raise RuntimeError(f'Failed to grant roles/storage.objectViewer to {connection_sa}:\n{r_iam.stderr}')
print('Granted roles/storage.objectViewer')

time.sleep(60)  # initial wait -- the query below retries further, since propagation is variable

Uploaded 6 objects to gs://statmike-mlops-349915/bq_ml/image/


Connection service account: bqcx-1026793852137-rjwp@gcp-sa-bigquery-condel.iam.gserviceaccount.com


Granted roles/storage.objectViewer


In [25]:
query = f"""
CREATE OR REPLACE EXTERNAL TABLE `{PROJECT_ID}.{DATASET_ID}.image_object_table`
WITH CONNECTION `{PROJECT_ID}.{LOCATION}.{CONNECTION_ID}`
OPTIONS (
  object_metadata = 'SIMPLE',
  uris = ['gs://{BUCKET}/{prefix}*']
)
"""
client.query(query).result()

# IAM propagation is genuinely variable -- retry on a permission error rather
# than gambling on the fixed wait above.
from google.api_core.exceptions import Forbidden

listing = f"""
SELECT REGEXP_EXTRACT(uri, r'[^/]+$') AS file, content_type, size, updated
FROM `{PROJECT_ID}.{DATASET_ID}.image_object_table`
ORDER BY file
"""
for attempt in range(1, 9):
    try:
        display(client.query(listing, job_config=no_cache).to_dataframe())
        break
    except (BadRequest, Forbidden) as e:
        if 'permission' not in str(e).lower() or attempt == 8:
            raise
        print(f'Attempt {attempt}/8: permission not yet propagated, waiting 20s...')
        time.sleep(20)

Attempt 1/8: permission not yet propagated, waiting 20s...


Attempt 2/8: permission not yet propagated, waiting 20s...


,file,content_type,size,updated
0,alpha.png,image/png,74,2026-09-11 17:26:14.814000+00:00
1,gradient.png,image/png,81,2026-09-11 17:26:14.684000+00:00
2,gray.png,image/png,68,2026-09-11 17:26:14.950000+00:00
3,swatch.jpg,image/jpeg,652,2026-09-11 17:26:15.151000+00:00
4,swatch.png,image/png,72,2026-09-11 17:26:14.555000+00:00
5,white.png,image/png,69,2026-09-11 17:26:15.041000+00:00


An object table exposes metadata columns (`uri`, `content_type`, `size`, `updated`) plus the `data` pseudocolumn holding the bytes. `data` is what `ML.DECODE_IMAGE` consumes.

Note that `content_type` is what Cloud Storage was told at upload time, not what the file is. It is metadata, not validation — filter on it if you like, but Step 2's decode error is the only thing that actually checks.

Now the whole pipeline in one statement: bytes → decode → fixed-size resize → grayscale → integers, written to a table rather than returned to the editor.

In [26]:
query = f"""
CREATE OR REPLACE TABLE `{PROJECT_ID}.{DATASET_ID}.image_features` AS
SELECT
  REGEXP_EXTRACT(uri, r'[^/]+$') AS file,
  ML.CONVERT_IMAGE_TYPE(
    ML.CONVERT_COLOR_SPACE(
      ML.RESIZE_IMAGE(ML.DECODE_IMAGE(data), 4, 4, FALSE),
      'GRAYSCALE')) AS features
FROM `{PROJECT_ID}.{DATASET_ID}.image_object_table`
WHERE content_type IN ('image/png', 'image/jpeg')
"""
client.query(query).result()

summary = f"""
SELECT file,
       features.dimensions      AS dimensions,
       ARRAY_LENGTH(features.values) AS n_values,
       features.values[OFFSET(0)]    AS first_value
FROM `{PROJECT_ID}.{DATASET_ID}.image_features`
ORDER BY file
"""
display(client.query(summary, job_config=no_cache).to_dataframe())

,file,dimensions,n_values,first_value
0,alpha.png,"[4, 4, 1]",16,124
1,gradient.png,"[4, 4, 1]",16,19
2,gray.png,"[4, 4, 1]",16,124
3,swatch.jpg,"[4, 4, 1]",16,122
4,swatch.png,"[4, 4, 1]",16,124
5,white.png,"[4, 4, 1]",16,255


Six source images of four different shapes and two formats, all reduced to the same dimensions and the same value count — a uniform feature vector a vision model can take as a batch. That uniformity is the entire point of the family, and it comes from two of the rules above: a `FALSE` aspect ratio pins height and width to exactly what was asked for, and the channel count is fixed by the last function in the chain (`GRAYSCALE` here, so 1 channel, per Step 5) rather than by the source images, which arrived with 1, 3, and 4 channels between them.

> **Write to a table, don't select the struct.** The reference pages warn that referencing these functions directly can fail to display in the BigQuery editor. That is a display limit, not a data limit — these images are tiny, but a 4000×3000 photo decodes to 36 million doubles, and the resize belongs *inside* the same statement that reads `data` so the large intermediate never materializes as a result.

The two remaining size limits are worth stating together, because one behaves differently than its wording suggests: **object-table image files must be under 20 MB and the `data` value passed to `ML.DECODE_IMAGE` must be at most 10 MB** — those are hard. The decoded `STRUCT`'s 60 MB ceiling is not: an image whose decoded struct would exceed it is **automatically downscaled, preserving aspect ratio**, rather than rejected. A silently resized input is a much subtler problem than a failed query, because the model still gets a valid tensor. If your pipeline handles large images, pin the size with an explicit `ML.RESIZE_IMAGE` instead of letting the ceiling do it for you.

### Where this goes next

These features feed a model that BigQuery did not train. `CREATE MODEL ... OPTIONS(model_type = 'TENSORFLOW', model_path = 'gs://...')` imports a vision SavedModel, and `ML.PREDICT` scores the table above — see [`models/imported/`](../../models/imported/) for the import mechanics end to end (its examples are tabular, but the `CREATE MODEL` shape is identical). Put the four functions in a `TRANSFORM` clause instead and the preprocessing is stored with the model and re-applied automatically at prediction time, which removes the possibility of training/serving skew in this stage of the pipeline entirely.


---
## Examples — `%%bigquery` Magics

The same pipeline using IPython magic commands — write SQL directly in cells without Python string wrapping.

In [27]:
%%bigquery --project {PROJECT_ID}

SELECT
  REGEXP_EXTRACT(uri, r'[^/]+$')                                          AS file,
  ML.DECODE_IMAGE(data).dimensions                                        AS source_dimensions,
  ML.RESIZE_IMAGE(ML.DECODE_IMAGE(data), 4, 4, FALSE).dimensions          AS resized_dimensions,
  ML.CONVERT_IMAGE_TYPE(ML.DECODE_IMAGE(data)).values[OFFSET(0)]          AS first_pixel_red
FROM `statmike-mlops-349915.bq_ml.image_object_table`
ORDER BY file

Query is running:   0%|          |

Downloading:   0%|          |

,file,source_dimensions,resized_dimensions,first_pixel_red
0,alpha.png,"[1, 2, 3]","[4, 4, 3]",200
1,gradient.png,"[4, 8, 3]","[4, 4, 3]",0
2,gray.png,"[1, 2, 3]","[4, 4, 3]",124
3,swatch.jpg,"[1, 2, 3]","[4, 4, 3]",105
4,swatch.png,"[1, 2, 3]","[4, 4, 3]",200
5,white.png,"[1, 1, 3]","[4, 4, 3]",255



---
## Examples — BigFrames

There is **no** `bigframes.ml` wrapper for any of the four. BigFrames has a multimodal `blob` accessor (`Series.blob.image_resize()`, `.image_normalize()`) that reads and writes image *files* through Cloud Storage — a different mechanism producing files, not the numeric `STRUCT` these functions produce. To reach the BigQuery functions themselves, run the SQL through `read_gbq`.

The struct comes back as a Python `dict`, so `dimensions` and `values` are ordinary keys.

In [28]:
import bigframes.pandas as bpd

bpd.close_session()  # Reset session to apply project/location settings
bpd.options.bigquery.project = PROJECT_ID
bpd.options.bigquery.location = LOCATION

bdf = bpd.read_gbq(f"""
SELECT REGEXP_EXTRACT(uri, r'[^/]+$') AS file,
       ML.RESIZE_IMAGE(ML.DECODE_IMAGE(data), 2, 2, FALSE) AS image
FROM `{PROJECT_ID}.{DATASET_ID}.image_object_table`
ORDER BY file
""")
pdf = bdf.to_pandas()
print(type(pdf['image'].iloc[0]))
print('dimensions:', pdf['image'].iloc[0]['dimensions'])
bdf

<class 'dict'>
dimensions: [2, 2, 3]


,file,image
0,alpha.png,"{'dimensions': array([2, 2, 3]), 'values': arr..."
1,gradient.png,"{'dimensions': array([2, 2, 3]), 'values': arr..."
2,gray.png,"{'dimensions': array([2, 2, 3]), 'values': arr..."
3,swatch.jpg,"{'dimensions': array([2, 2, 3]), 'values': arr..."
4,swatch.png,"{'dimensions': array([2, 2, 3]), 'values': arr..."
5,white.png,"{'dimensions': array([2, 2, 3]), 'values': arr..."



---
## Cleanup

Drop the tables, the connection, and the uploaded objects — then delete the temporary reservation, which is the one resource here that could keep costing money. The shared dataset and the bucket itself are left alone.

In [29]:
for table in ['image_features', 'image_object_table']:
    client.query(f'DROP TABLE IF EXISTS `{PROJECT_ID}.{DATASET_ID}.{table}`').result()
    print(f'Dropped {table}')

subprocess.run(
    ['bq', 'rm', '-f', '--connection', '--project_id', PROJECT_ID, '--location', LOCATION, CONNECTION_ID],
    capture_output=True, text=True,
)
print(f'Deleted connection {CONNECTION_ID}')

deleted = 0
for blob in gcs_bucket.list_blobs(prefix=prefix):
    blob.delete()
    deleted += 1
print(f'Deleted {deleted} GCS object(s) under gs://{BUCKET}/{prefix}')

Dropped image_features


Dropped image_object_table


Deleted connection bq_ml_image_demo


Deleted 6 GCS object(s) under gs://statmike-mlops-349915/bq_ml/image/


In [30]:
# Delete the temporary reservation and its assignment (safe to re-run — ignores NotFound)
from google.api_core.exceptions import NotFound

try:
    for existing_assignment in reservation_client.list_assignments(parent=reservation.name):
        reservation_client.delete_assignment(name=existing_assignment.name)
    reservation_client.delete_reservation(name=reservation.name)
    print(f'Deleted reservation {reservation.name} and its assignment(s)')
except NotFound:
    print('Reservation already deleted')

print('Remaining reservations in this location:',
      [r.name for r in reservation_client.list_reservations(parent=reservation_parent)] or 'none')

Deleted reservation projects/statmike-mlops-349915/locations/US/reservations/bqml-image-temp and its assignment(s)
Remaining reservations in this location: none



### Remove all shared resources

When finished with **all** notebooks, uncomment and run the cell below to delete the shared dataset and everything in it.

In [31]:
# client.delete_dataset(
#     f'{PROJECT_ID}.{DATASET_ID}',
#     delete_contents=True,
#     not_found_ok=True
# )
# print(f'Dataset {PROJECT_ID}.{DATASET_ID} deleted')